### ISSUE TITLE
Race condition when starting multiple terminals? pyenv: cannot rehash: .pyenv-shim exists

### ISSUE THREAD

**Issue by fsc-eriker** [pyenv/pyenv#2829]
This morning when I logged in after an OS update (MacOS Sonoma 14.1) I got `-bash: python: command not found` in all my restarted terminals.

In one of them, I could also see

```
pyenv: cannot rehash: /Users/fsc-eriker/.pyenv/shims/.pyenv-shim exists
```

Of course, the file no longer existed by the time I checked.

Running `pyenv rehash` in one of the terminals resolved the problem in all of them.

My suspicion is that there is a race condition when many terminals start up in quick succession after a reboot, though I have been running `pyenv` for several years and this is the first time I see this problem.

### Prerequisite
* [X] Make sure your problem is not listed in [the common build problems](https://github.com/pyenv/pyenv/wiki/Common-build-problems).
* [X] Make sure no duplicated issue has already been reported in [the pyenv issues](https://github.com/pyenv/pyenv/issues). You should look for closed issues, too.
* [X] Make sure you are not asking us to help solving your specific issue.
  * GitHub issues is opened mainly for development purposes. If you want to ask someone to help solving your problem, go to some community site like [Gitter](https://gitter.im/yyuu/pyenv), [StackOverflow](https://stackoverflow.com/questions/tagged/pyenv), etc.
* [ ] Make sure your problem is not derived from packaging (e.g. [Homebrew](https://brew.sh)).
  * Please refer to the package documentation for the installation issues, etc.
* [ ] Make sure your problem is not derived from plugins.
  * This repository is maintaining `pyenv` and the default `python-build` plugin only. Please refrain from reporting issues of other plugins here.

### Description
- [X] Platform information: MacOS 14.1
- [X] OS architecture (e.g. amd64): Intel
- [X] pyenv version: 2.3.17
- [X] Python version: 3.11.3 (among others)
- [ ] C Compiler information (e.g. gcc 7.3):  clang I guess? I have 15.0.0
- [ ] Please attach the debug trace of the failing command as a gist:
  * Run `env PYENV_DEBUG=1 <faulty command> 2>&1 | tee trace.log` and attach `trace.log`. E.g. if you have a problem with installing Python, run `env PYENV_DEBUG=1 pyenv install -v <version> 2>&1 | tee trace.log` (note the `-v` option to `pyenv install`).
- [ ] If you have a problem with installing Python, please also attach `config.log` from the build directory
  * The build directory is reported after the "BUILD FAILED" message and is usually under `/tmp`.
- [ ] If the build succeeds but the problem is still with the build process (e.g. the resulting Python is missing a feature), please attach
  * the debug trace from reinstalling the faulty version with `env PYENV_DEBUG=1 pyenv install -f -k -v <version> 2>&1 | tee trace.log`
  * `config.log` from the build directory. When using `pyenv install` with `-k` as per above, the build directory will be under `$PYENV_ROOT/sources`.

I have `pyenv-virtualenv` plugin but I imagine it is not involved in this issue. Unfortunately, because there is no way to repro reliably, I cannot rule out that this is caused by Homebrew or the plugin or etc.

Several old issues seem related but too old to be directly relevant as duplicates; #1092, #1145, maybe #1137 and #1747.

**Comment by native-api** [comment]
> pyenv: cannot rehash: /Users/fsc-eriker/.pyenv/shims/.pyenv-shim exists

> My suspicion is that there is a race condition when many terminals start up in quick succession after a reboot

That's exactly it.

Does that cause a problem though? All terminal instances should be fine if at least one rehash has succeeded.

**Comment by native-api** [comment]
> because there is no way to repro reliably

`touch "${PYENV_ROOT}/shims/.pyenv-shim"`

That's what one of our tests uses to model this situation.

**Comment by fsc-eriker** [comment]
> Does that cause a problem though? All terminal instances should be fine if at least one rehash has succeeded.

Indeed, the ticket is about a case where this did not happen, i.e. none of the terminal instances was fine (as far as I was able to check; I did not individually review more than 5-6)

**Comment by fsc-eriker** [comment]
As a tentative workaround I changed my `.bashrc` to run `eval "$(pyenv init - --no-rehash)"` and instead set up a separate login item to run `eval "$(pyenv init -)"` just once. I don't think I can know whether it works until much later.

**Comment by native-api** [comment]
I cannot reproduce the problem. `pyenv-rehash` is designed for this situation: it uses a lock file and waits with a 60 second timeout if it's already held before bailing out.
Moreover, the stock startup code silences error messages from `pyenv rehash`. So you may have local customizations in place.

Make sure you have the latest, stock version of Pyenv.
If it's still a problem, please provide the debug trace as per the issue template -- of both shell startup and a standalone `pyenv rehash`.

**Comment by rsantellan** [comment]
After installing on LMDE 6 (Debian 12) I'm having the same problem, on startup guake is started an when I try to open a terminal it shows on guake:

`pyenv: cannot rehash: ~/.pyenv/shims/.pyenv-shim exists`

Is pretty common issue for me (I turn off the laptop every day and work mainly with a terminal). 

How can I provide more information to this issue ?

**Comment by native-api** [comment]
@rsantellan Use the issue template to provide the information necessary to diagnose a problem. It appears when you create a new issue.

**Comment by techwizrd** [comment]
I’m seeing what looks like the same failure mode on macOS with a newer pyenv, but with a more visible shell-startup impact.

Environment:
- macOS
- fish
- pyenv 2.6.26
- installed via Homebrew

What happens:
- starting a new interactive fish shell was taking about 60 seconds
- the slowdown came from `pyenv init` during shell startup
- `pyenv init` emits `command pyenv rehash`
- `pyenv rehash` then waits the full 60 second timeout trying to acquire the shim lock/prototype file

Observed error:

```text
pyenv: cannot rehash: couldn't acquire lock ~/.pyenv/shims/.pyenv-shim for 60 seconds. Last error message:
pyenv-rehash: line 22: ~/.pyenv/shims/.pyenv-shim: cannot overwrite existing file
```

In my case, the file was still present after the failure:

```text
~/.pyenv/shims/.pyenv-shim
```

This looks like a stale/orphaned lock/prototype file case, not a sandbox writability issue.

Measurements:
- `fish --no-config -i -c exit`: about `0.01s`
- normal `fish -i -c exit` before fix: about `60.6s`
- after removing the stale `.pyenv-shim` and changing shell init to `pyenv init - --no-rehash fish | source`: about `0.28s` to `0.34s`

So there seem to be two separate concerns:
1. a stale `.pyenv-shim` can block `pyenv rehash` for the full timeout
2. unconditional `rehash` in `pyenv init` makes that failure hit every new shell

I realize `--no-rehash` is a workable user-side fix, but I wanted to add evidence that the `.pyenv-shim exists` problem can still translate into a consistent 60-second shell startup delay when the lock file is left behind.

If useful, I can provide more details from local testing.

**Comment by native-api** [comment]
I guess we can assume that a lockfile older than, say, 10 minutes is stale.

We still don't know how this happens but this will at least unblock affected users.
